# Test data acknowledgements

`loop.gfa` was created by Shaun Jackman.
[Here](https://github.com/sjackman/assembly-graph/blob/fef9fada23ddfb3da04db8221fac1ca8c99bfc66/loop.gfa)
is a link to the version of the file used.

`sample1.gfa` and `sample2.gfa` were also created by Shaun Jackman. We
downloaded them from the [gfalint](https://github.com/sjackman/gfalint) repository
(in the `examples/` folder).

`all_line_types.gfa1.gfa` and `all_line_types.gfa2.gfa` are from
[the Gfapy repository's test data](https://github.com/ggonnella/gfapy/tree/master/tests/testdata).
Furthermore, `path_of_edges.gfa` is a modified version of `all_line_types.gfa2.gfa`.

`marygold_fig2a.gml` is based on Figure 2a from
[Nijkamp et al. 2013](https://www.ncbi.nlm.nih.gov/pubmed/24058058).

`E_coli_LastGraph` is a modified version of one of
[Bandage's](http://rrwick.github.io/Bandage/)
sample graphs, which apparently was created from data located on NCBI
[here](https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=272917).
We modified the version included here so that every node's sequence has
just been changed to `ACGT` -- this is because we don't really care about the
actual sequence data for the tests here so much as the topology of the graph
(so there's no reason to store like 9 MB of sequence data).

`flye_yeast.gv` is a _S. cerevisiae_ assembly graph generated by Flye. We
downloaded it [from AGB's GitHub repository](https://github.com/almiheenko/AGB/tree/master/test_data/flye_yeast).
`flye_yeast_assembly_info.txt` was downloaded from the same repository.
(`flye_yeast.agp` is just a file I made up for testing.)

`sheepgut_g1217.gfa` corresponds to a small, highly-mutated region of a MAG
"smoothed-read" assembly graph generated from the sheep gut dataset used in
[Bickhart and Kolmogorov et al. 2022](https://www.nature.com/articles/s41587-021-01130-z).
This MAG (and the process of read smoothing) is discussed in
[Fedarko et al. 2022](https://genome.cshlp.org/content/32/11-12/2119).

`aug1_subgraph.gml` is a small subgraph of a stool sample (SRS049959)'s
assembly graph, which was scaffolded with MetaCarvel.

`chr21mat_subgraph_1.gv` and `chr21mat_subgraph_2.gv` are small subgraphs (with
some slight modifications) of de Bruijn graphs produced by
[jumboDBG](https://github.com/AntonBankevich/LJA/)
from the maternal haplotype of chromosome 21 (from
[Jarvis & Formenti et al. 2022](https://www.nature.com/articles/s41586-022-05325-5)).
Technically these two files are derived from different de Bruijn graphs (I used two different versions of jumboDBG to create them, if I remember correctly?),
but the details aren't very relevant for these files' test purposes. The main
reason I use these small graphs is that they triggered bugs in MetagenomeScope, so
now they're immortalized here as regression tests.
(Thank you to Zhenmiao Zhang for downloading the chromosome 21 haplotype to our server :)

Similarly, `chr15_subgraph.gv` is a subgraph of a de Bruijn graph produced by
jumboDBG from the chromosome 15 haplome from version 2.0 of the
[CHM13](https://github.com/marbl/CHM13) assembly.

