# Copyright (C) 2016-- Marcus Fedarko, Jay Ghurye, Todd Treangen, Mihai Pop
# Authored by Marcus Fedarko
#
# This file is part of MetagenomeScope.
#
# MetagenomeScope is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# MetagenomeScope is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with MetagenomeScope.  If not, see <http://www.gnu.org/licenses/>.
####
# Defines various configurable variables for the python script.

# The conversion factor between points (in GraphViz output) and inches, as used
# by GraphViz. GraphViz uses 72 points per inch, but here we use 63
# points per inch -- this is kind of a compromise. In old versions of
# MetagenomeScope, at times we would pass inch values (for node / pattern
# widths and heights) directly to the JS interface and have that code do the
# conversion using an INCHES_TO_PIXELS value of 54 -- which was inconsistent,
# since things like edge control points were just stored as points and not
# converted to pixels.
#
# We now do all the conversions in the Python code, but
# I've noticed that using the 72 points-per-inch value for nodes kinda squishes
# the graph too much, making things look gross. (Some of the edge arrows for
# bubbles don't even show up in the interface.) Using a smaller value helps a
# bit -- using 54 for this spaces out the graph a bit too much, and using the
# midpoint between 54 and 72 (63) seems ok. Since this really scales _all_
# distances in the graph, I think, changing this value doesn't change the
# intepretation of the graph -- at least for node sizes, it seems to rescale
# all nodes. But I should really try to formally look into this sometime...
POINTS_PER_INCH = 63.0

# The base we use when logarithmically scaling contig dimensions from length
NODE_SCALING_LOG_BASE = 10
# The minimum/maximum area of a node.
# These variables are used directly in GraphViz, so they're technically in
# "inches". That being said, "inches" are really an intermediate unit from our
# perspective since (at least as of writing) they're converted to points and
# then passed directly to Cytoscape.js. See POINTS_PER_INCH above for details.
MAX_NODE_AREA = 10
MIN_NODE_AREA = 1
NODE_AREA_RANGE = MAX_NODE_AREA - MIN_NODE_AREA
# Proportions of the "long side" of a contig, for various levels of contigs in
# the graph.
# Used for the lower 25% (from 0% to 25%) of contigs in a component
LOW_LONGSIDE_PROPORTION = 0.5
# Used for the middle 50% (from 25% to 75%) of contigs in a component
MID_LONGSIDE_PROPORTION = 2.0 / 3.0
# Used for the upper 25% (from 75% to 100%) of contigs in a component
HIGH_LONGSIDE_PROPORTION = 5.0 / 6.0

# Factor by which collapsed node groups' dimensions are scaled
# As with the dimensions of Node objects in the preprocessing script, these
# will be switched in the viewer interface (due to the graph being rotated)
# -- so _W_ refers to the vertical dimension of the node group, and _H_
# refers to the horizontal dimension.
COLL_CL_W_FAC = 1.0 / 2.0
COLL_CL_H_FAC = 1.0 / 2.0

### Frequently-used Graphviz settings ###
# More info on these available at www.graphviz.org/doc/info/attrs.html

### Global graph settings (applied to every node/edge/etc. in the graph) ###
#
# NOTE that most of the current settings here are important for keeping the
# drawings generated by MetagenomeScope "consistent," so changing these around
# could result in strange-looking output.
#
# General graph style. Feel free to insert newlines/tabs for readability.
# Leaving this empty is also fine, if you just want default graph-wide settings
# Note that we rely on having a rankdir value of "TB" for the entire graph,
# in order to facilitate rotation in the .xdot viewer. (So I'd recommend not
# changing that unless you have a good reason.)
# Any text here must end without a semicolon.
GRAPH_STYLE = "rankdir=LR"

# Style applied to every node in the graph.
#
# Note that this style also impacts node groups in the "backfilled" layouts
# generated using -pg or -px, since node groups are temporarily represented as
# nodes.
#
# "fixedsize=true" is needed in order to prevent node labels from causing nodes
# to be resized. Might trigger some warnings from Graphviz but oh well, better
# that than nodes getting assigned the wrong sizes.
#
# "orientation=90" is needed in order to get nodes to be correctly
# rotated (to point in the left --> right direction), which makes the graph
# look as intended when rankdir=LR. Shoutouts to
# https://stackoverflow.com/a/45431027 for the wisdom that the node
# "orientation" attr existed.
GLOBALNODE_STYLE = "fixedsize=true,orientation=90"

# Whether or not to color in nodes in the drawings produced by dot (without
# having an effect on the viewer interface, since it doesn't care what colors
# dot says a node/node group is).
# Since this will impact node groups in backfilled layouts, if you use this
# style it's recommended that you also set COLOR_PATTERNS to True. (Otherwise,
# the backfilled layouts' drawings generated by dot will color all node groups
# #888888.)
#
# NOTE that COLOR_NODES and COLOR_PATTERNS are both set to True in order to
# generate Figs. 3 and 4 in the MetagenomeScope paper, so these config options
# should be left in this file so those figures can be replicated.
COLOR_NODES = True
# The actual style addition is kept separate from COLOR_NODES to make it easy
# to toggle this behavior and reference the toggling of this behavior
if COLOR_NODES:
    GLOBALNODE_STYLE += ',style=filled,fillcolor="#888888"'

# Style applied to every edge in the graph.
# Keeping headport=w,tailport=e is strongly recommended (assuming we keep
# rankdir=LR), since that impacts how edges are positioned and drawn in the
# graph (and the JS code expects edges to originate from nodes' "headports"
# and end at nodes' "tailports").
GLOBALEDGE_STYLE = "headport=w,tailport=e"

# Style applied (directly) to every cluster in the graph.
# Keeping margin=0 is strongly recommended, since otherwise cluster bounding
# boxes can take up extra space in the graph (the resulting drawings would also
# not mesh well with the viewer interface, where compound nodes are drawn with
# a zero margin).
GLOBALCLUSTER_STYLE = "margin=0"

### Other misc. config variables ###
# Used to create lines in logging output like =====
SEPBIG = "="
SEPSML = "-"

# The amount of indentation to use for each level in DOT files. You could also
# use a \t character here, if you're the sort of person who prefers tabs ._.
INDENT = "  "

# Used as the suffixes of split nodes.
SPLIT_SEP = "-"
SPLIT_LEFT = "L"
SPLIT_RIGHT = "R"
SPLIT_LEFT_SUFFIX = SPLIT_SEP + SPLIT_LEFT
SPLIT_RIGHT_SUFFIX = SPLIT_SEP + SPLIT_RIGHT

# This is mostly all configurable, but the JS callbacks for searching/etc
# rely on these being "-L" and "-R". That could be made more elegant by like
# passing the suffixes to the function but that seems not worth it
# Also, name_utils.get_splitname_base() assumes that these both have a length
# of 2. seriously don't change these lol
if SPLIT_LEFT_SUFFIX != "-L" or SPLIT_RIGHT_SUFFIX != "-R":
    raise ValueError(
        "Hey, you can't change that without updating the JS searching code."
    )

# Pattern types -- used internally.
PT_BUBBLE = 0
PT_CHAIN = 1
PT_CYCLICCHAIN = 2
PT_FRAYEDROPE = 3
PT_BIPARTITE = 4

# Maps pattern types to human-readable names.
PT2HR = {
    PT_BUBBLE: "Bubble",
    PT_CHAIN: "Chain",
    PT_CYCLICCHAIN: "Cyclic Chain",
    PT_FRAYEDROPE: "Frayed Rope",
    PT_BIPARTITE: "Bipartite",
}

# Maps pattern types to human-readable names without spaces. For those silly
# cases where we want to include pattern names in other IDs (e.g. in the names
# of clusters in exported DOT files)...
PT2HR_NOSPACE = {
    PT_BUBBLE: "bubble",
    PT_CHAIN: "chain",
    PT_CYCLICCHAIN: "cyclicchain",
    PT_FRAYEDROPE: "frayedrope",
    PT_BIPARTITE: "bipartite",
}

# Whether or not to specify colors for node groups in .gv/.xdot files. If this
# is True, then PT2COLOR is used to set the colors.
COLOR_PATTERNS = True

PT2COLOR = {
    # matches "cornflowerblue" in graphviz
    PT_BUBBLE: "#9abaf3",

    # matches "green2" in graphviz
    PT_FRAYEDROPE: "#59f459",

    # this one used to match "salmon" in graphviz (#fcaca3 i think?) and then i
    # changed it to be less flashy since there are a lot of chains involved
    # in hierarchical decomp stuff and i want people to take this seriously
    PT_CHAIN: "#eedddd",

    # matches "darkgoldenrod1" in graphviz
    PT_CYCLICCHAIN: "#ffd163",

    PT_BIPARTITE: "#c8a4ed",
}

# Draw types. Used internally for communicating between flush() and draw().
DRAW_ALL = "all"
DRAW_CCS = "ccs"
DRAW_AROUND = "around"

# Some context on port numbers, for anyone who stumbles upon this in the future
# (including future me):
#
# - ModDotPlot, another bio tool that is set up as a Dash app, does not
#   explicitly set a min/max port number --
#   https://github.com/marbl/ModDotPlot/blob/d7675f99536df74b95efcca56e20b09bf42e432d/src/moddotplot/moddotplot.py#L144-L149
#
# - Dash itself requires that the port is an int in the range [1, 65535] --
#   https://github.com/plotly/dash/blob/30afe785c26c2bb6f4dc56efd0c9001d4384acfc/dash/dash.py#L2428
#
# - https://linux.die.net/man/5/services says that 'Port numbers below 1024
#   (so-called "low numbered" ports) can only be bound to by root'.
#
# - https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers lists a
#   bunch of stuff about port numbers, and is how I ran into the "services" man
#   page cited above.
#
# - If you try running dash with a port number of 1023, my system complains
#   about not having permission. However, using a port number of 1024 works!
#
# So ... this all seems to suggest to me that a minimum of >= 1024 should be
# okay. And I guess we should impose a maximum of 65535 in order to
# "fail fast," rather than waiting for Dash to throw the error (after we
# spend a bunch of time processing the graph).
#
# I am sure this will not be perfect but it doesn't have to be -- this is
# gonna have to be up to the user to figure out if they don't want to use 8050
# (or whatever the default is).
MIN_PORT = 1024
MAX_PORT = 65535
