# Copyright (C) 2016-- Marcus Fedarko, Jay Ghurye, Todd Treangen, Mihai Pop
# Authored by Marcus Fedarko
#
# This file is part of MetagenomeScope.
#
# MetagenomeScope is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# MetagenomeScope is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with MetagenomeScope.  If not, see <http://www.gnu.org/licenses/>.
####
# Defines various configurable variables for the python script.

# The conversion factor between points (in GraphViz output) and inches, as used
# by GraphViz. GraphViz uses 72 points per inch, but here we use 63
# points per inch -- this is kind of a compromise. In old versions of
# MetagenomeScope, at times we would pass inch values (for node / pattern
# widths and heights) directly to the JS interface and have that code do the
# conversion using an INCHES_TO_PIXELS value of 54 -- which was inconsistent,
# since things like edge control points were just stored as points and not
# converted to pixels.
#
# We now do all the conversions in the Python code, but
# I've noticed that using the 72 points-per-inch value for nodes kinda squishes
# the graph too much, making things look gross. (Some of the edge arrows for
# bubbles don't even show up in the interface.) Using a smaller value helps a
# bit -- using 54 for this spaces out the graph a bit too much, and using the
# midpoint between 54 and 72 (63) seems ok. Since this really scales _all_
# distances in the graph, I think, changing this value doesn't change the
# intepretation of the graph -- at least for node sizes, it seems to rescale
# all nodes. But I should really try to formally look into this sometime...
POINTS_PER_INCH = 63.0

# The base we use when logarithmically scaling contig dimensions from length
NODE_SCALING_LOG_BASE = 10
# The minimum/maximum area of a node.
# These variables are used directly in GraphViz, so they're technically in
# "inches". That being said, "inches" are really an intermediate unit from our
# perspective since (at least as of writing) they're converted to points and
# then passed directly to Cytoscape.js. See POINTS_PER_INCH above for details.
MAX_NODE_AREA = 10
MIN_NODE_AREA = 1
NODE_AREA_RANGE = MAX_NODE_AREA - MIN_NODE_AREA
# Proportions of the "long side" of a contig, for various levels of contigs in
# the graph.
# Used for the lower 25% (from 0% to 25%) of contigs in a component
LOW_LONGSIDE_PROPORTION = 0.5
# Used for the middle 50% (from 25% to 75%) of contigs in a component
MID_LONGSIDE_PROPORTION = 2.0 / 3.0
# Used for the upper 25% (from 75% to 100%) of contigs in a component
HIGH_LONGSIDE_PROPORTION = 5.0 / 6.0

# Factor by which collapsed node groups' dimensions are scaled
# As with the dimensions of Node objects in the preprocessing script, these
# will be switched in the viewer interface (due to the graph being rotated)
# -- so _W_ refers to the vertical dimension of the node group, and _H_
# refers to the horizontal dimension.
COLL_CL_W_FAC = 1.0 / 2.0
COLL_CL_H_FAC = 1.0 / 2.0

### Frequently-used Graphviz settings ###
# More info on these available at www.graphviz.org/doc/info/attrs.html

### Global graph settings (applied to every node/edge/etc. in the graph) ###
#
# NOTE that most of the current settings here are important for keeping the
# drawings generated by MetagenomeScope "consistent," so changing these around
# could result in strange-looking output.
#
# General graph style. Feel free to insert newlines/tabs for readability.
# Leaving this empty is also fine, if you just want default graph-wide settings
# Note that we rely on having a rankdir value of "TB" for the entire graph,
# in order to facilitate rotation in the .xdot viewer. (So I'd recommend not
# changing that unless you have a good reason.)
# Any text here must end without a semicolon.
GRAPH_STYLE = "rankdir=LR"

# Style applied to every node in the graph.
#
# Note that this style also impacts node groups in the "backfilled" layouts
# generated using -pg or -px, since node groups are temporarily represented as
# nodes.
#
# "fixedsize=true" is needed in order to prevent node labels from causing nodes
# to be resized. Might trigger some warnings from Graphviz but oh well, better
# that than nodes getting assigned the wrong sizes.
#
# "orientation=90" is needed in order to get nodes to be correctly
# rotated (to point in the left --> right direction), which makes the graph
# look as intended when rankdir=LR. Shoutouts to
# https://stackoverflow.com/a/45431027 for the wisdom that the node
# "orientation" attr existed.
GLOBALNODE_STYLE = "fixedsize=true,orientation=90"

# Whether or not to color in nodes in the drawings produced by dot (without
# having an effect on the viewer interface, since it doesn't care what colors
# dot says a node/node group is).
# Since this will impact node groups in backfilled layouts, if you use this
# style it's recommended that you also set COLOR_PATTERNS to True. (Otherwise,
# the backfilled layouts' drawings generated by dot will color all node groups
# #888888.)
#
# NOTE that COLOR_NODES and COLOR_PATTERNS are both set to True in order to
# generate Figs. 3 and 4 in the MetagenomeScope paper, so these config options
# should be left in this file so those figures can be replicated.
COLOR_NODES = True
# The actual style addition is kept separate from COLOR_NODES to make it easy
# to toggle this behavior and reference the toggling of this behavior
if COLOR_NODES:
    GLOBALNODE_STYLE += ',style=filled,fillcolor="#888888"'

# Style applied to every edge in the graph.
# Keeping headport=w,tailport=e is strongly recommended (assuming we keep
# rankdir=LR), since that impacts how edges are positioned and drawn in the
# graph (and the JS code expects edges to originate from nodes' "headports"
# and end at nodes' "tailports").
GLOBALEDGE_STYLE = "headport=w,tailport=e"

# Style applied (directly) to every cluster in the graph.
# Keeping margin=0 is strongly recommended, since otherwise cluster bounding
# boxes can take up extra space in the graph (the resulting drawings would also
# not mesh well with the viewer interface, where compound nodes are drawn with
# a zero margin).
GLOBALCLUSTER_STYLE = "margin=0"

### Other misc. config variables ###
# The default values for -maxn and -maxe in the collate script.
MAXN_DEFAULT = 7999
MAXE_DEFAULT = 7999

# Used to add an extra line below the first logging message, and just before
# the last logging message.
SEPARATOR_CHAR = "-"

# The amount of indentation to use for each level in DOT files. You could also
# use a \t character here, if you're the sort of person who prefers tabs ._.
INDENT = "  "

# Maps "internal" names for node / edge properties to human-readable ones.
# This is passed on to the JavaScript code, so that the UI there can use these
# human-readable names.
# (As we add on more filetype parsers, feel free to extend these structures.)
NODEATTR2HR = {
    "name": "ID",
    "length": "Length",
    "orientation": "Orientation",
    "depth": "Coverage",
    "cov": "Coverage",
    "gc": "GC Content",
    "gc_content": "GC Content",
}
EDGEATTR2HR = {
    # Ideally we'd be more precise about what exactly these are the mean and
    # standard deviation of, but i am not completely sure. I think they
    # represent the "mean and standard deviation for the implied distance
    # between a pair of contigs," to quote the MetaCarvel paper? But they could
    # also represent "the library size (mean and standard deviation)," maybe,
    # so for the sake of avoiding being wrong I am being conservative for now
    "mean": "Mean",
    "stdev": "Standard Deviation",
    "bsize": "Bundle Size",
    "multiplicity": "Multiplicity",
    "id": "ID",
    "approx_length": "Approximate Length",
    "cov": "Coverage",
    "color": "Color",
    "first_nt": "First Nucleotide",
    "kmer_cov": "k-mer Coverage",
}

# Used as the suffixes of split nodes.
SPLIT_SEP = "-"
SPLIT_LEFT = "L"
SPLIT_RIGHT = "R"

# Pattern types -- used internally.
PT_BUBBLE = 0
PT_CHAIN = 1
PT_CYCLICCHAIN = 2
PT_FRAYEDROPE = 3

# Maps pattern types to human-readable names.
PT2HR = {
    PT_BUBBLE: "Bubble",
    PT_CHAIN: "Chain",
    PT_CYCLICCHAIN: "Cyclic Chain",
    PT_FRAYEDROPE: "Frayed Rope",
}

# Maps pattern types to human-readable names without spaces. For those silly
# cases where we want to include pattern names in other IDs (e.g. in the names
# of clusters in exported DOT files)...
PT2HR_NOSPACE = {
    PT_BUBBLE: "Bubble",
    PT_CHAIN: "Chain",
    PT_CYCLICCHAIN: "CyclicChain",
    PT_FRAYEDROPE: "FrayedRope",
}

# Whether or not to specify colors for node groups in .gv/.xdot files. If this
# is True, then PT2COLOR is used to set the colors.
COLOR_PATTERNS = True

# Maps pattern types to colors.
PT2COLOR = {
    PT_BUBBLE: "#9abaf3",
    PT_FRAYEDROPE: "#59f459",
    PT_CHAIN: "#eedddd",
    PT_CYCLICCHAIN: "#ffd163",
}
